### ISSUE TITLE
Add a permission check CommandElement

### ISSUE THREAD

**Issue by boformer** [SpongePowered/SpongeAPI#636]
The permission check element can be wrapped around a ``CommandElement`` (like  ``GenericArguments.optional(..)``).

It only passes the argument to the wrapped element if the command sender has (or does not have) a specified permission:
```java
CommandElement permissionTest(CommandElement element, String permission, boolean value);
```
The element is especially useful to generate usage strings when the player is not allowed to use all features of a command.

Usage Example: A teleport command that supports teleporting **yourself** or (with an extra permission) **other players** (``/tp [otherPlayer] <targetPlayer>``). Another extra permission enables a ``-s = silent`` flag that hides the "xxx teleported to you" message.

Players without the extra permission should only see this usage string: ``/tp <targetPlayer>``.
Admins should see the usage string: ``/tp [otherPlayer] <targetPlayer>``:
```java
CommandSpec.builder()
.setArguments(
    GenericArguments.permissionTest(
        GenericArguments.player(Texts.of("otherPlayer")), 
        "teleport.other", true),
    GenericArguments.player(Texts.of("targetPlayer")))
...
```

There should be a similar mechanism for command flags. (maybe a ``flagWithPermission`` method in the flag builder)

What do you think, @zml2008 ?

**Comment by ZephireNZ** [comment]
+1 from me

**Comment by zml2008** [comment]
I could see this -- it's growing on me more as I write this comment. I like how it feels as a solution to having certain arguments require additional permissions, though it also has the limitation of not being more contextual -- so if the other player you specified was yourself, you would still be checked for the .other permission.

Another thing your example brings up (which imo is a bigger concern): with how your arguments are currently ordered, the argument handling would always take a player as the otherPlayer, and throw an exception if the user doesn't provide a targetPlayer, meaning that the otherPlayer is not truly optional. I'm not currently sure how that would best be resolved though & don't have a ton of free time in the next few days, so anybody feel free to take a stab at that issue.

Also the fact that you don't static-import methods in GenericArguments will bug me forever, though that's more of a personal codestyle thing.

**Comment by MattBDev** [comment]
I could really use this. My plugin has a lot of command arguments which require their own permissions. Even the command arguments of those commands have their own permissions.

**Comment by boformer** [comment]
@zml2008 Do you have a good idea how this could work with flags?

Maybe like this?

```java
flags()
    .permission("myplugin.flag.a")
    .flag("a")
    .permission("myplugin.flag.b-and-q")
    .flag("b")
    .valueFlag(integer(t("quot")), "q")
    .permission(null)
    .flag("c")
    .buildWith(...)
```
Require "myplugin.flag.a" for flag a,
require "myplugin.flag.b-and-q" for flag b and q,
require no permission for flag c.

**Comment by zml2008** [comment]
this would be easy for value-flags -- just use the permission commandelement with the flag. For non-value flags, it probably makes the most sense to add a permissionFlag(String perm, String... flagSpecs) method -- that leaves things a bit less ambiguous imo.

**Comment by zml2008** [comment]
So accidentally tagged the wrong issue, but this was fixed in 8ab104f00966a97fb1b3c634d20e6e927bdaa598
